### PR TITLE
feature_metadata_hierarchy

### PR DESCRIPTION
**Pull request by Samulus**
Tested: 
- no legacy hierarchy regressions `find output/ -name '*.b3dm' | xargs md5sum` 
- no `undefined` or `NaN` keys in the generated `output/Hierarchy/*.gltf|*.glb`

### CODE REVIEW

File: samples-generator/lib/createBatchTableHierarchy.ts [hunk 1]
```diff
@@ -142,16 +160,65 @@ function createBatchTableHierarchy(options) {
...
+
+            if (!useGlb) {
+                return Promise.all([
+                    saveJson(tilePath, result, options.prettyJson, gzip),
+                    saveJson(tilesetJsonPath, tilesetJson, options.prettyJson, gzip)
+                ]);
+            } else {
+                return Promise.all([
+                    gltfToGlb(result, gltfConversionOptions).then(function(out) {
+                        saveBinary(tilePath, out.glb, gzip);
```
Comment (lilleyse) [inline samples-generator/lib/createBatchTableHierarchy.ts]: Is `return` still needed?

```suggestion
                        return saveBinary(tilePath, out.glb, gzip);
```

File: samples-generator/lib/createBatchTableHierarchy.ts [hunk 2]
```diff
@@ -142,16 +160,65 @@ function createBatchTableHierarchy(options) {
             }
         }
         var batchedMesh = Mesh.batch(clonedMeshes);
+
+        if (use3dTilesNext || useGlb) {
```
Comment (lilleyse) [inline samples-generator/lib/createBatchTableHierarchy.ts]: Is `use3dTilesNext` sufficient?

```suggestion
        if (use3dTilesNext) {
```

File: samples-generator/lib/createBatchTableHierarchy.ts [hunk 3]
```diff
@@ -142,16 +160,65 @@ function createBatchTableHierarchy(options) {
...
+                mesh: batchedMesh
+            });
+        }
+
         return createGlb({
             mesh : batchedMesh
         });
-    }).then(function(glb) {
+    }).then(function(result) {
+        if (use3dTilesNext || useGlb) {
```
Comment (lilleyse) [inline samples-generator/lib/createBatchTableHierarchy.ts]: Same comment

```suggestion
        if (use3dTilesNext) {
```

File: samples-generator/lib/createFeatureHierarchySubExtension.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+ * @param gltf glTF asset to edit. This glTF asset should already have a
+ * `CESIUM_3dtiles_feature_metadata` extension in the extensions member of
+ * the glTF asset.
+ * @param classes An array of FeatureHierarchyClasses to insert into the glTF
+ * asset
+ * @param classIds classIds for the feature hierarchy
+ * @param instanceCount Total number of instances. Should be equal to the sum
+ * of each classes[i].instancesCount
+ * @param [parentIds] Optional list of parentIds associated with each classIds
+ * @param [parentCounts] Optional numerical array containing
```
Comment (lilleyse) [inline samples-generator/lib/createFeatureHierarchySubExtension.ts]: containing...?

File: samples-generator/bin/3d-tiles-samples-generator.ts [hunk 5]
```diff
@@ -5,13 +5,14 @@ var Cesium = require('cesium');
 var fsExtra = require('fs-extra');
```
Comment (lilleyse) [inline samples-generator/bin/3d-tiles-samples-generator.ts]: `CESIUM_3dtiles_feature_hierarchy` should be listed in `extensionsUsed`

Comment (lilleyse) [reply samples-generator/bin/3d-tiles-samples-generator.ts]:
```
          "attributes": {
            "POSITION": 0,
            "NORMAL": 1,
            "_BATCHID": 2
          },
```

The `use3dTilesNext` flag must not be getting propagated correctly to `createGltf`. `_BATCHID` should be `_FEATURE_ID_0`.

File: samples-generator/bin/3d-tiles-samples-generator.ts [hunk 6]
```diff
@@ -1188,7 +1195,9 @@ function createHierarchyLegacy() {
         transform : buildingsTransform,
```
Comment (lilleyse) [inline samples-generator/bin/3d-tiles-samples-generator.ts]: I would just skip `BatchTableHierarchyLegacy` in `--3d-tiles-next` mode.

File: samples-generator/bin/3d-tiles-samples-generator.ts (hunk 5 above)
Comment (lilleyse) [reply samples-generator/bin/3d-tiles-samples-generator.ts]: The hierarchy extension should be contained in an individual feature table, like

```
  "extensions": {
    "CESIUM_3dtiles_feature_metadata": {
      "featureTables": [
        {
          "properties": {
            "height": {
              "accessor": 4
            },
            "area": {
              "accessor": 5
            }
          },
          "featureCount": 30,
          "extensions": {
            "CESIUM_3dtiles_feature_hierarchy": {
            }
          }
        }
      ]
    }
  }
}
```

**Comment by Samulus** [comment]
@lilleyse all comments addressed

**Comment by lilleyse** [comment]
Great, thanks @Samulus!

**Pull request merged**